### PR TITLE
[python-package] Add support for NumPy 2.0, test against nightly versions of dependencies (fixes #6454)

### LINKED ISSUE #6454: [python-package] NumPy 2.0 support
## Summary

`numpy` 2.0 is coming out soon, and has release candidates up.

We should test `lightgbm` against it, and if any changes are required, try to get them into the next release (#6439).

## Motivation

`numpy` is a hard runtime dependency of `lightgbm`. It would not be a good experience for users if incompatibility with its new major version leads to needing to manually pin `numpy<2` in their environments.

## Description

At a minimum, we should manually test `lightgbm` against  the `numpy` release candidate. See https://github.com/dmlc/treelite/issues/560 for a migration guide.

It also would be great to set up CI that runs against nightlies of `numpy`, `pandas`, `scikit-learn`, and `scipy`, maybe following `xarray`'s example: https://github.com/pydata/xarray/blob/31111b3afe44fd6f7dac363264e94186cc5168d2/.github/workflows/upstream-dev-ci.yaml

## References

Links to other projects' upgrade and testing efforts: https://github.com/numpy/numpy/issues/26191

### PR DESCRIPTION
**Pull request by jameslamb**
Fixes #6454.

This PR makes **`lightgbm` compatible with `numpy` 2.x**

* converts `np.array(..., copy=False)` calls (which are no longer supported in NumPy 2.x) with `np.asarray(...)`.
* adds a CI job testing `lightgbm` against nightlies of `matplotlib`, `numpy`, `pandas`, `pyarrow`, `scipy`, and `scikit-learn`

Also fixes some other smalls things I noticed thanks to the new CI job.

* makes it clearer when `test_arrow.py` tests are failing because `pyarrow` is installed but `cffi` is not
* skips on test on `pandas>=3.0` (which is not release yet), to deal with a change to `pd.DataFrame(...)`'s copying behavior when built from a `numpy` array

### What does it mean that `np.array(..., copy=False)` is "no longer supported"?

Raises an exception if you try to ask it to do an impossible cast.

```python
import numpy as np

X = np.array([1, 2, 3, 4], dtype=np.float32)
np.array(X, dtype=np.float64, copy=False)
```

```text
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: Unable to avoid copy while creating an array as requested.
If using `np.array(obj, copy=False)` replace it with `np.asarray(obj)` to allow a copy when needed (no behavior change in NumPy 1.x).
For more details, see https://numpy.org/devdocs/numpy_2_0_migration_guide.html#adapting-to-changes-in-the-copy-keyword.
```

### So `lightgbm` is going to start creating unnecessary copies again on `pandas` input, if you use `numpy` 2.0`?

No.

**pandas** is going to start creating copies on NumPy input under some conditions. Follow the conversation at https://github.com/pandas-dev/pandas/issues/58913.

cc @jmoralez, since you did so much work on this e.g. in #4927 and #5612

### CODE REVIEW

File: tests/python_package_test/test_arrow.py [hunk 1]
```diff
@@ -20,6 +20,10 @@
 else:
     import pyarrow as pa  # type: ignore
 
+    assert (
+        lgb.compat.PYARROW_INSTALLED is True
+    ), "'pyarrow' and its dependencies must be installed to run the arrow tests"
```
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:25]: I found myself in a situation where `pyarrow` was installed but `cffi` was not.

That cause this import to fail:

https://github.com/microsoft/LightGBM/blob/e0cda880fc74ca6d1b7d6cb425a24e3a69764bb1/python-package/lightgbm/compat.py#L212

which led to all of the `pyarrow` imports being defined with the placeholders like this:

https://github.com/microsoft/LightGBM/blob/e0cda880fc74ca6d1b7d6cb425a24e3a69764bb1/python-package/lightgbm/compat.py#L227-L231

which led to every Arrow test failing in a hard-to-understand way.

```shell
pytest tests/python_package_test/test_arrow.py
```

```text
==== short test summary info ====
FAILED tests/python_package_test/test_arrow.py::test_dataset_construct_fuzzy[<lambda>-dataset_params0] - AssertionError: assert False
... 
FAILED tests/python_package_test/test_arrow.py::test_predict_ranking - TypeError: Wrong type(ChunkedArray) for label.
==== 139 failed in 11.77s ====
```

As of this change, it'll fail with a slightly clearer error message (and earlier, without running all the tests).

```text
E   AssertionError: 'pyarrow' and its dependencies must be installed to run the arrow tests
```

File: .github/workflows/python_package.yml [hunk 2]
```diff
@@ -89,6 +116,7 @@ jobs:
         run: |
           docker run \
             --rm \
+            --env CMAKE_BUILD_PARALLEL_LEVEL=${{ env.CMAKE_BUILD_PARALLEL_LEVEL }} \
```
Comment (jameslamb) [inline .github/workflows/python_package.yml:119]: This is one that I missed in #6458, noticed it when replicating some of the configuration for the new CI job here

**Review by guolinke — APPROVED** [review]
Thank you!

**Review by jmoralez — APPROVED** [review]

**Comment by jameslamb** [comment]
Thanks very much for the reviews! I'm very excited to now have this CI job testing against nightlies of `lightgbm`'s dependencies. I hope it won't be too noisy, and will help us catch issues before any `lightgbm` users do.

**Pull request merged**